### PR TITLE
Update feature

### PR DESCRIPTION
**Pull request by NZT48**
Changes:
- Support for hardhat as local network
- Update asset state
- Get latest and latest finalized asset state
- Wait for asset update finalization
- Expand storing period for asset
- Add tokens to asset

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,7 +32,7 @@
     "dependencies": {
         "assertion-tools": "^2.0.2",
         "axios": "^0.27.2",
-        "dkg-evm-module": "^3.2.1",
+        "dkg-evm-module": "file:../dkg-evm-module",
```
Comment (NZT48) [inline package.json:35]: To be changed later

File: managers/asset-operations-manager.js [hunk 2]
```diff
@@ -398,60 +431,331 @@ class AssetOperationsManager {
...
+                agreementId,
+                blockchain,
+            );
+
+            const now = await this.blockchainService.getBlockchainTimestamp(blockchain);
+            const currentEpoch = Math.floor(
+                (now - agreementData.startTime) / agreementData.epochLength,
+            );
+
+            const epochsLeft = agreementData.epochsNumber - currentEpoch;
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:584]: State can be finalized in the current epoch and we should always have a buffer here (include current epoch) imo. Let's check if we have the same mechanism on the node side (for bid validation)

File: services/blockchain-service/blockchain-service-base.js [hunk 3]
```diff
@@ -213,10 +248,83 @@ class BlockchainServiceBase {
...
+
     async getAssetOwner(tokenId, blockchain) {
         return this.callContractFunction('ContentAssetStorage', 'ownerOf', [tokenId], blockchain);
     }
 
+    async burnAsset(tokenId, blockchain) {
+        return this.executeContractFunction('ContentAsset', 'burnAsset', [tokenId], blockchain);
+    }
+
+    async extendAssetStoringPeriod(tokenId, epochsNumber, tokenAmount, blockchain) {
```
Comment (u-hubar) [inline services/blockchain-service/blockchain-service-base.js:268]: Missing allowance increase for ServiceAgreementV1

File: services/blockchain-service/blockchain-service-base.js [hunk 4]
```diff
@@ -213,10 +248,83 @@ class BlockchainServiceBase {
...
+        return this.executeContractFunction(
+            'ContentAsset',
+            'updateAssetStoringPeriod',
+            [tokenId, epochsNumber, tokenAmount],
+            blockchain,
+        );
+    }
+
+    async addTokens(tokenId, tokenAmount, blockchain) {
```
Comment (u-hubar) [inline services/blockchain-service/blockchain-service-base.js:277]: Missing allowance increase for ServiceAgreementV1

File: managers/asset-operations-manager.js [hunk 5]
```diff
@@ -478,6 +789,58 @@ class AssetOperationsManager {
...
+            epochsNumber,
+            tokenAmount,
+            blockchain,
+        );
+
+        const { tokenId } = resolveUAL(UAL);
+        await this.blockchainService.extendAssetStoringPeriod(
+            tokenId,
+            epochsNumber,
+            tokenAmount,
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:821]: I guess we need to use bid suggestion here as well?

File: .eslintrc.js [hunk 6]
```diff
@@ -11,7 +11,7 @@ module.exports = {
     },
     rules: {
         'linebreak-style': ['error', 'unix'],
-        'camelcase': 0,
+        camelcase: 0,
```
[the remark this answers is not in the document]
Comment (zeroxbt) [reply .eslintrc.js:15]: probably prettier changed it. it is the same as before

**Pull request merged**